onedal/datatypes/data_conversion_sua_iface.cpp

```diff
@@ -0,0 +1,221 @@
+/*******************************************************************************
+* Copyright 2024 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifdef ONEDAL_DATA_PARALLEL
+#define NO_IMPORT_ARRAY
+
+#include <stdexcept>
+#include <string>
+#include <utility>
+
+#include "oneapi/dal/common.hpp"
+#include "oneapi/dal/detail/common.hpp"
+#include "oneapi/dal/table/homogen.hpp"
+#include "oneapi/dal/table/detail/homogen_utils.hpp"
+
+#include "onedal/common/policy_common.hpp"
+#include "onedal/datatypes/data_conversion_sua_iface.hpp"
+#include "onedal/datatypes/utils/dtype_conversions.hpp"
+#include "onedal/datatypes/utils/dtype_dispatcher.hpp"
+#include "onedal/datatypes/utils/sua_iface_helpers.hpp"
+
+namespace oneapi::dal::python {
+
+// Please follow <https://intelpython.github.io/dpctl/latest/
+// api_reference/dpctl/sycl_usm_array_interface.html#sycl-usm-array-interface-attribute>
+// for the description of `__sycl_usm_array_interface__` protocol.
+
+// Convert python object to oneDAL homogen table with zero-copy by use
+// of `__sycl_usm_array_interface__` protocol.
+template <typename Type>
+dal::table convert_to_homogen_impl(py::object obj) {
```

Comment:
As we only currently support homogen tables in oneDAL, this will have to be updated for when heterogeneous table support comes, both on input and output of oneDAL.  (I see it also in the npy_strides function).

Reply:
I have already documented here which kind of tables are supported.